gsoc/gsoc2016/ideas_yt.md

```diff
@@ -0,0 +1,166 @@
+  gracefully fall back to using pyplot instead of the object oriented API.
+
+* A YTEP describing the proposed approach for modifying yt's plotting
+  infrastructure to support matplotlib's interactive plotting backends.
+
+* The implementation for the YTEP submitted as a bitbucket pull request to the
+  main yt repository.
+
+### Testing improvements
+
+*Suggested Mentor(s):* [Kacper Kowalik](https://bitbucket.org/xarthisius),
+ [Nathan Goldbaum](https://bitbucket.org/ngoldbaum)
+
+*Difficulty:* Beginner to Advanced, depending on where the student takes the
+ project
+
+*Knowledge needed*: Familiarity with the `nose` testing package.
+
+*Programming skills:* Python, Cython
+
+#### Description
+
+Currently yt's test suite is split between unit tests (which take about 45
+minutes to run) and answer tests, which are normally only run on a continuous
+integration server. Altogether the tests only cover about a third of the yt
+codebase, so much of the code in yt needs test coverage.  Additionally, the
+tests take a long time to run, and we would like to reduce the test runtime
+while simultaneously increasing code coverage.
+
+This project could go in a number of directions:
+
+* Implement a way to retrofit the current tests for different geometries
+  (e.g. cartesian, cylindrical, and spherical coordinates) and data styles
+  (e.g. particle data, as well as various kind of mesh data, including uniform
+  resolution, octree, patch AMR, and unstructured meshes). Ideally this would
+  allow us to test all functionality for all possible data styles. This will
+  require learning and improving the "Stream" frontend, which allows the
+  injestion of in-memory data into yt.
+
+* Identify areas of the code that are not well tested and devise tests for
+  them. This will require measuring the test coverage of yt's Python and
+  Cython components. The student working on this will need to gain familiarity
+  with untested or undertested parts of the codebase and add new
+  tests. Optimally the new tests will make use of new reusable infrastructure
+  that will be helpful for tests across the yt codebase.
+
+* Improve volume rendering and visualization unit tests. Right now visualization
+  tests rely heavily on answer testing and image comparison. It would be more
+  flexible and easier to understand when things go wrong if the tests instead
+  compared with a predicted answer using some sort of simplified geometry or via
+  introspection.
+
+*Deliverables*:
+
+* Develop a framework for measuring test covering in yt's python and cython
+  components. Triage the reports to look for areas that are user facing and have
+  poor test coverage.
+
+* Make a number of pull requests adding tests across the yt codebase.
+
+* Modify existing testing infrastructure or develop new test infrastructure to
+  improve testing of yt functionality on different data types.
+
+### Domain contexts and domain-specific fields
```

Comment:
Oh, and to be clear: in my experience this "wide-open" problem is bigger for the *mentors* than the students.  That is, having a wide-open project often means the mentor has to spend a lot more time hand-holding and guiding the student than they would on something more focused.  If the mentors are OK with that then that's fine, but it's probably even odds that it's more of a time sink than just doing it yourself might be... (again, depends on the student, but you might not know that until they've been selected if it's someone who's new to your community)